clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -475,20 +475,25 @@ Sender newSender(LogContext logContext, KafkaClient kafkaClient, ProducerMetadat
                 apiVersions);
     }
 
+    private static int lingerMs(ProducerConfig config) {
+        return (int) Math.min(config.getLong(ProducerConfig.LINGER_MS_CONFIG), Integer.MAX_VALUE);
+    }
+
     private static int configureDeliveryTimeout(ProducerConfig config, Logger log) {
         int deliveryTimeoutMs = config.getInt(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG);
-        int lingerMs = config.getInt(ProducerConfig.LINGER_MS_CONFIG);
+        int lingerMs = lingerMs(config);
         int requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+        int lingerAndRequestTimeoutMs = (int) Math.min((long) lingerMs + requestTimeoutMs, Integer.MAX_VALUE);
 
-        if (deliveryTimeoutMs < Integer.MAX_VALUE && deliveryTimeoutMs < lingerMs + requestTimeoutMs) {
+        if (deliveryTimeoutMs < Integer.MAX_VALUE && deliveryTimeoutMs < lingerAndRequestTimeoutMs) {
```

Comment:
It seems like we didn't enforce this too well if `lingerMs + requestTimeoutMs` overflowed. It would have been possible to misconfigure the producer with specifying lower `deliveryTimeoutMs` and higher `lingerMs + requestTimeoutMs`. We will now raise an exception if that is the case.

The test fixes needed in this patch are for the same reason, where `lingerMs` was set to `Integer.MAX_VALUE` and `deliveryTimeoutMs` was set to the default value of 2 minutes.